### ISSUE TITLE
ROWNUM transformation doesn't work in the presence of DISTINCT

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11661]
The `ROWNUM` to `LIMIT` transformation doesn't seem to work in the presence of `DISTINCT`:

```sql
select distinct a
from t
where rownum <= 10
```

Remove `distinct`, and it works:

```sql
select A
from T
fetch next 10 rows only
```

**Comment by wal1dcha1b** [comment]
Yeah, that's true, in fact i have many requests that didin't get parsed in the presence of distinct or unique

**Comment by lukaseder** [comment]
Thanks for the feedback. Notice that `GROUP BY` is also a known limitation: https://github.com/jOOQ/jOOQ/issues/10716

**Comment by lukaseder** [comment]
The difficulty of the transformation is that the `ROWNUM` filter is applied *before* `DISTINCT`, logically. So, if a query without `DISTINCT` yields 10 rows:

```sql
select a from t where rownum <= 10
```

Then, it might well be that a query *with* `DISTINCT` will yield less than 10 rows:

```sql
select distinct a from t where rownum <= 10
```

This is almost always a bug in the source SQL. I doubt that you would ever want to have an "at most 10" rows semantics, if there could be more than 10. For example:

```sql
WITH t (a) AS (SELECT MOD(LEVEL, 2) FROM dual CONNECT BY LEVEL <= 20 UNION ALL SELECT 3 FROM dual)
SELECT a FROM t WHERE rownum <= 10;
```

Yields

```
|A  |
|---|
|1  |
|0  |
|1  |
|0  |
|1  |
|0  |
|1  |
|0  |
|1  |
|0  |
```

Whereas

```sql
WITH t (a) AS (SELECT MOD(LEVEL, 2) FROM dual CONNECT BY LEVEL <= 20 UNION ALL SELECT 3 FROM dual)
SELECT DISTINCT a FROM t WHERE rownum <= 10;
```

Yields

```
|A  |
|---|
|1  |
|0  |
```

You'd expect the value `3` to be present in the result, because you want less than 10 rows. But it's not there, because the `ROWNUM` filter is applied non-deterministically on the CTE *before* removing duplicates. Akin to the following Java stream usages:

```sql
// Wrong
stream.limit(10).distinct();

// Intended, correct
stream.distinct().limit(10);
```

We can, of course, still translate the queries as they were written (including the above bug).

**Comment by lukaseder** [comment]
... alternatively, if you're on Oracle 12c or more, you could re-write your query first to a more correct one using Oracle 12c and standard SQL `FETCH FIRST n ROWS ONLY`, which can be translated easily by jOOQ.

**Comment by wal1dcha1b** [comment]
Unfortuanetly, we have too much bad written requests, we will have to re-write it in a proper way

**Comment by lukaseder** [comment]
Well, the fix will be included, and will also fix `GROUP BY` and possibly a few other limitations, to try to emulate the same behaviour again as in Oracle.

For now, the fix will introduce a new regression in 3.15.0 which should be fixed by the time it is released: We're now using derived tables to make sure the `ROWNUM` filter applies *before* anything else (e.g. *before* `GROUP BY` or `DISTINCT`). As such, if the projection (`SELECT` clause) contains duplicate column names, that will now fail, because that's not allowed in derived tables. We'll have to apply the usual column renaming ceremony. That will be done via #10716

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.15.0. The fix relies on numerous new SQL transformation features available in 3.15 only, which is why it won't be backported to 3.14.
